File: matlab/grad2D.m
function [dudx,dvdy] = grad2D(Uo,Vo,dx,dy,n,m,upbound,dobound,...
    lebound,ribound)
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% This function computes the terms du/dx, dv/dy with O(h^2) centered
% differences (if posible)
%
% Poisson equation
%
%   dP2/d2x + dP2/d2y = -(rho/dt)*(du*/dx + dv*/dy)
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Entries
%   Uo, vector with velocities in X direction
%   Uv, vector with velocities in Y direction
%   dx, differential in X
%   dy, differential in Y
%   n, number of nodes in X direction
%   m, number of nodes in Y direction
%   upbound, indices of the top boundary (ub)
%   dobound, indices of the bottom boundary (db)
%   lebound, indices of the left boundary (lb)
%   ribound, indices of the right boundary (rb)
%      *[upbound, dobound, lebound, ribound] € bound
%   boundin, internal boundaries
%
% Grid and what bound, upboundin, doboundin, leboundin, and riboundin means
%
% Enumartion used in the grid
%
%    25   26   27   28   29   30
%
%    19   20   21   22   23   24
%
%    13   14   15   16   17   18
%
%    7    8    9    10   11   12
%
%    1    2    3    4    5    6
%
% upbound(ub), dobound(db), lebound(lb), ribound(rb), boundint(bi)
% internal nodes(i)
%
%    ulb  ub   ub   ub   ub   urb
%
%    lb   i    i    i    i    rb
%
%    lb   i    i    i    i    rb
%
%    lb   i    i    i    i    rb
%
%    dlb  db   db   db   db   drb
%
% dul --> top-left node
% ddl --> bottom-left node
% ddr --> bottom-right node
% dur --> top-right node
%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
   
    %Allocating memory for the result
    dudx = zeros(n*m,1);
    dvdy = zeros(n*m,1);
    
    %Loop
    for i=1:n*m
        %
        % Derivatives in the corners
        %
        if i==1
            dudx(i) = (1/dx)*(Uo(i+1) - Uo(i));
            dvdy(i) = (1/dy)*(Vo(i+n) - Vo(i));
        elseif i == n
            dudx(i) = (1/dx)*(Uo(i) - Uo(i-1));
            dvdy(i) = (1/dy)*(Vo(2*i) - Vo(i));
        elseif i == n*m-n+1
            dudx(n*m-n+1) = (1/dx)*(Uo(n*m-n+2) - Uo(n*m-n+1));
            dvdy(n*m-n+1) = (1/dy)*(Vo(n*m-n+1) - Vo(n*m-2*n+1));
        elseif i == n*m
            dudx(n*m) = (1/dx)*(Uo(n*m) - Uo(n*m-1));
            dvdy(n*m) = (1/dy)*(Vo(n*m) - Vo(n*m-n));
        %
        % Derivatives in the boundaries
        %
        elseif ismember(i,upbound)
            dudx(i) = (1/(2*dx))*(Uo(i+1) - Uo(i-1));
            dvdy(i) = (1/dy)*(Vo(i) - Vo(i-n));
        elseif ismember(i,dobound)
            dudx(i) = (1/(2*dx))*(Uo(i+1) - Uo(i-1));
            dvdy(i) = (1/dy)*(Vo(i+n) - Vo(i));
        elseif ismember(i,lebound)
            dudx(i) = (1/dx)*(Uo(i+1) - Uo(i));
            dvdy(i) = (1/(2*dy))*(Vo(i+n) - Vo(i-n));
        elseif ismember(i,ribound)
            dudx(i) = (1/dx)*(Uo(i) - Uo(i-1));
            dvdy(i) = (1/(2*dy))*(Vo(i+n) - Vo(i-n));
        %
        % Derivatives in the internal portion
        %
        else
            dudx(i) = (1/(2*dx))*(Uo(i+1) - Uo(i-1));
            dvdy(i) = (1/(2*dy))*(Vo(i+n) - Vo(i-n));
        end
    end

end

File: src/grad2d.f
!                    *****************
                     SUBROUTINE GRAD2D
!                    *****************
     & (DUDX,DVDY,UO,VO,UPBOUND,DOBOUND,LEBOUND,RIBOUND)

!
!***********************************************************************
! 2D-NAVIER STOKES SOLVER - FINITE DIFFERENCES
!***********************************************************************
!
!BRIEF    1) COMPUTE DIFFUSION COEFFICIENTS AND MATRIX.
!
!HISTORY  SERGIO CASTIBLANCO
!+        25/01/2021
!+        TRANSLATION FOR ORIGINAL MATLAB IMPLEMENTATION
!
!!
!!
!! POISSON EQUATION
!!
!!   DP2/D2X + DP2/D2Y = -(RHO/DT)*(DU*/DX + DV*/DY)
!!
!! GRID AND WHAT BOUND, UPBOUNDIN, DOBOUNDIN, LEBOUNDIN, AND RIBOUNDIN
!! MEANS
!!
!! ENUMARTION USED IN THE GRID
!!
!!    25   26   27   28   29   30
!!
!!    19   20   21   22   23   24
!!
!!    13   14   15   16   17   18
!!
!!    7    8    9    10   11   12
!!
!!    1    2    3    4    5    6
!!
!! UPBOUND(UB), DOBOUND(DB), LEBOUND(LB), RIBOUND(RB), BOUNDINT(BI)
!! INTERNAL NODES(I)
!!
!!    ULB  UB   UB   UB   UB   URB
!!
!!    LB   I    I    I    I    RB
!!
!!    LB   I    I    I    I    RB
!!
!!    LB   I    I    I    I    RB
!!
!!    DLB  DB   DB   DB   DB   DRB
!!
!! DUL --> TOP-LEFT NODE
!! DDL --> BOTTOM-LEFT NODE
!! DDR --> BOTTOM-RIGHT NODE
!! DUR --> TOP-RIGHT NODE
!!
!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
!| DUDX      |<--| D(UO)/DX VECTOR                                     |
!| DVDY      |<--| D(VO)/DY VECTOR                                     |
!| UO        |-->| UO                                                  |
!| VO        |-->| VO                                                  |
!| UPBOUND   |-->| TOP BOUNDARY INDICES                                |
!| DOBOUND   |-->| BOTTOM BOUNDARY INDICES                             |
!| LEBOUND   |-->| LEFT BOUNDARY INDICES                               |
!| RIBOUND   |-->| RIGHT BOUNDARY INDICES                              |
!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
!
      USE DECLARATIONS_NUMERICAL, ONLY: NX,NY,DX,DY
!
      IMPLICIT NONE
!
!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
!
      DOUBLE PRECISION,DIMENSION(NX*NY),INTENT(IN) :: UO,VO
      DOUBLE PRECISION,DIMENSION(NX*NY),INTENT(OUT) :: DUDX,DVDY
      INTEGER, DIMENSION(NX-2), INTENT(IN) :: UPBOUND, DOBOUND
      INTEGER, DIMENSION(NY-2), INTENT(IN) :: LEBOUND, RIBOUND
!
!  IN SUBROUTINE VARIABLES
!
      INTEGER :: I
!
!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
!
      DO I = 1,NX*NY
!
!  DERIVATIVES IN THE CORNERS
!
        IF(I.EQ.1) THEN
          DUDX(I) = (1/DX)*(UO(I+1) - UO(I))
          DVDY(I) = (1/DY)*(VO(I+NX) - VO(I))
        ELSEIF(I.EQ.NX) THEN
          DUDX(I) = (1/DX)*(UO(I) - UO(I-1))
          DVDY(I) = (1/DY)*(VO(2*I) - VO(I))
        ELSEIF(I.EQ.(NX*NY-NX+1)) THEN
          DUDX(I) = (1/DX)*(UO(NX*NY-NX+2) - UO(NX*NY-NX+1))
          DVDY(I) = (1/DY)*(VO(NX*NY-NX+1) - VO(NX*NY-2*NX+1))
        ELSEIF(I.EQ.(NX*NY)) THEN
          DUDX(I) = (1/DX)*(UO(NX*NY) - UO(NX*NY-1))
          DVDY(I) = (1/DY)*(VO(NX*NY) - VO(NX*NY-NX))
!
!  DERIVATIVES IN THE BOUNDARIES
!
        ELSEIF(ANY(I.EQ.UPBOUND)) THEN
          DUDX(I) = (1/(2*DX))*(UO(I+1) - UO(I-1));
          DVDY(I) = (1/DY)*(VO(I) - VO(I-NX));
        ELSEIF(ANY(I.EQ.DOBOUND)) THEN
          DUDX(I) = (1/(2*DX))*(UO(I+1) - UO(I-1));
          DVDY(I) = (1/DY)*(VO(I+NX) - VO(I));
        ELSEIF(ANY(I.EQ.LEBOUND)) THEN
          DUDX(I) = (1/DX)*(UO(I+1) - UO(I));
          DVDY(I) = (1/(2*DY))*(VO(I+NX) - VO(I-NX));
        ELSEIF(ANY(I.EQ.RIBOUND)) THEN
          DUDX(I) = (1/DX)*(UO(I) - UO(I-1));
          DVDY(I) = (1/(2*DY))*(VO(I+NX) - VO(I-NX));
!
!  DERIVATIVES IN THE INTERNAL NODES
!
        ELSE
          DUDX(I) = (1/(2*DX))*(UO(I+1) - UO(I-1));
          DVDY(I) = (1/(2*DY))*(VO(I+NX) - VO(I-NX));
        ENDIF
      ENDDO



!
!~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
!
      END SUBROUTINE GRAD2D



